### ISSUE TITLE
Losing connection with vJoy

### ISSUE THREAD

**Issue by laqqah** [AndersMalmgren/FreePIE#164]
I have a .py script that reads my thrustmaster joystick buttons and creates new buttons in a vJoy device from combinations of those. This because the TM buttons have ON or OFF position and when in OFF I can't bind them in simulators as I would like. So I create a vJoy button only device which I feed to enable buttons when the joystick button is not detected. And then I can bind the vJoy buttons inside the simulator.

It all works nice for a while but FreePIE losing connection to vJoy requires alt+tab out of simulation to stop script (shift+f5) and restart it again (f5), which makes it work again.

Sometimes connection lost in 10 minutes sometimes it lasts up to two hours or more.

* I have checked vJoyList for owner of devices.
Sometimes the vJoy is still owned by FreePIE and sometimes the owner PID is blank.

* FreePIE's own monitor is working. I can see in "Watch" window that the values are changing so connection with joystick is working fine (even without restarting script).

* There is nothing written to the "Console" or "Error" windows when it fails

* vJoy is still running and a simple stopping the script in freepie and starting it again makes it work fine again. No restart of vJoy or computer needed.

* Checking the vJoy device through "USB game controllers" -> Vjoy -> Properties shows all darkened out buttons so none is received.

It appears to me that the connection between FreePIE and vJoy gets broken.

![image](https://user-images.githubusercontent.com/621386/46570008-351cf880-c95e-11e8-88b7-6e2d562d510d.png)

**Comment by laqqah** [comment]
This is the vJoy driver information. It is the one I get when I install "latest".
![image](https://user-images.githubusercontent.com/621386/46570275-f25d1f80-c961-11e8-820f-367ae909690b.png)

**Comment by MarijnS95** [comment]
Would it be possible to share your script so that we can have a look at what's going on?

If anything, FreePIE is just sending axis/button states to the vjoy library that communicates with the driver. Sounds as if FreePIE is not pushing updates anymore.

**Comment by laqqah** [comment]
Github does not support .py files. Do you want me to zip it up or a screenshot good enough ?

![image](https://user-images.githubusercontent.com/621386/46570601-f5a6da00-c966-11e8-9569-ce8a4a040cb0.png)

**Comment by laqqah** [comment]
I have FreePIE running right now with the vJoy unresponsive. No owner of vJoy device.
![image](https://user-images.githubusercontent.com/621386/46570700-abbef380-c968-11e8-8fde-122305c63101.png)

**Comment by laqqah** [comment]
I'm trying to befriend the FreePIE code and Visual Studio environment.
I managed to build and run and I'm trying out this code which I call from SetButton() for now.
I do not see any obvious place (other than inside every function) to place this checkalive() call.
For now I throw exception but will change to reconnect if this get caught.

Do you think this will be acceptable ?

        public void CheckAlive()
        {
            var status = joystick.GetVJDStatus(Index);
            if (status != VjdStat.VJD_STAT_OWN)
            {
                throw new Exception(string.Format("We no longer own the vJoy device"));
            }
        }

**Comment by AndersMalmgren** [comment]
Can't you check it in the BeforeNextExecute

**Comment by laqqah** [comment]
Seems like a better place. Yes, I missed that one.
I'll see if it catches the problem from there.

**Comment by MarijnS95** [comment]
You could do that in `DoBeforeNextExecute`, though I'd recommend printing it to the log instead of raising an exception.

At the same time you could add a conditional to all `joystick.Set*` calls; if it returns false, log it and reacquire the device, similar to the vJoy client linked above.

**Comment by laqqah** [comment]
I'll change to System.Console.WriteLine or to be quiet when I reaquire the device.
If that is the problem.
I hope that is the problem.
If that is not the problem then we have a problem :-)

**Comment by laqqah** [comment]
Well I'm having problems recreating the detach.
I disconnected the joystick on purpose to make sure there is no glitching.
When disconnected the monitoring stops working as well so now I know that is not the issue.

Will continue trying to recreate detach.

**Comment by laqqah** [comment]
I've finally managed to lose connection and verified FreePIE automatically connects back again.
Printing a line on console about "lost connection and attempting to reacquire."

I'll try to have a PR ready by tomorrow. Busy weekend.

**Comment by MarijnS95** [comment]
Cool :). Was it the `joystick.Set*` calls returning false after some time, or is it detected using `joystick.GetVJDStatus(Index)`?

**Comment by laqqah** [comment]
I check joystick.GetVJDStatus(Index) in CheckAlive() which is called from the DoBeforeNextExecute loop.

It might be less expensive to just check retval from joystick.Set* calls but I'm thinking at that time it is already too late and you will miss one event? So for now I poll at every loop.

Still no idea why it disconnects tho..
While trying to fix this there has been times when it went 8h+ without disconnect (I gave up and shut down) and there are times when disconnects happen within 15 minutes.

**Comment by MarijnS95** [comment]
I agree it is a weird situation, especially since no-one has ever faced this before. However, this looks like expected behaviour thanks to the boolean returns from the `Set*` functions so you shouldn't worry too much about it.

Using the boolean will be faster because it's readily available (unlike` GetVJDStatus` which has to call into the C library, which in turn does an `IoControl` call to the driver to retrieve the result). The difference will be negligible though, just like every python call to `setButton`, `setAxis` and others do exactly the same. (If anyone wants to tackle this: store the values in `JoystickState` and call `UpdateVJD` with it from `DoBeforeNextExecute`).

If false is returned and the event is missed, it can easily be sent again after reacquiring the device, in the same `if` block. Such an edgecase for every `set` method adds duplication which can be hidden with some lambda functions ;). I'll let @AndersMalmgren decide on that.

```C#
private void UpdateJoystickField(Action updater)
{
    if (!updater())
    {
        joystick.Acquire();
        if (!updater())
            throw new Exception("Failed to update joystick after .Acquire()!");
    }
}```

**Comment by laqqah** [comment]
PR created.
#165
